File: 2018/9.9jisuanke_xuzhou/B.cpp
#include<bits/stdc++.h>
using namespace std;
#define rep(i,a,b) for(int i=a;i<=b;i++)
#define per(i,a,b) for(int i=a;i>=b;i--)
#define clr(a,x) memset(a,x,sizeof(a))
const int MAXN=1e3+10;
int a[MAXN],b[MAXN],c[MAXN];
bool dp[MAXN][220];
int main(){
    int n,m,k,l;
    scanf("%d%d%d%d",&n,&m,&k,&l);
    rep(i,1,n){
        scanf("%d%d%d",&a[i],&b[i],&c[i]);
    }
    clr(dp,0);
    rep(i,k+100,200){
        dp[n+1][i]=1;
    }
    per(i,n,1){
        if(i%2==1){
            rep(j,0,200){
                bool f=0;
                if(a[i]>0){
                    int w=j+a[i];
                    if(w>200)w=200;
                    if(dp[i+1][w])f=1;
                }
                if(b[i]>0){
                    int w=j-b[i];
                    if(w<0)w=0;
                    if(dp[i+1][w])f=1;
                }
                if(c[i]==1){
                    int w=j-100;
                    w=-w;
                    w+=100;
                    if(dp[i+1][w])f=1;
                }
                if(f)dp[i][j]=1;
            }
        }else{
            rep(j,0,200){
                bool f=0;
                bool fz=1;
                if(a[i]>0){
                    int w=j+a[i];
                    if(w>200)w=200;
                    if(dp[i+1][w])f=1;
                    else fz=0;
                }
                if(b[i]>0){
                    int w=j-b[i];
                    if(w<0)w=0;
                    if(dp[i+1][w])f=1;
                    else fz=0;
                }
                if(c[i]==1){
                    int w=j-100;
                    w=-w;
                    w+=100;
                    if(dp[i+1][w])f=1;
                    else fz=0;
                }
                if(f&&fz)dp[i][j]=1;
            }
        }
    }
    if(dp[1][m+100]){
        puts("Good Ending");
        return 0;
    }
    clr(dp,0);
    rep(i,0,100+l){
        dp[n+1][i]=1;
    }
    per(i,n,1){
        if(i%2==0){
            rep(j,0,200){
                bool f=0;
                if(a[i]>0){
                    int w=j+a[i];
                    if(w>200)w=200;
                    if(dp[i+1][w])f=1;
                }
                if(b[i]>0){
                    int w=j-b[i];
                    if(w<0)w=0;
                    if(dp[i+1][w])f=1;
                }
                if(c[i]==1){
                    int w=j-100;
                    w=-w;
                    w+=100;
                    if(dp[i+1][w])f=1;
                }
                if(f)dp[i][j]=1;
            }
        }else{
            rep(j,0,200){
                bool f=0;
                bool fz=1;
                if(a[i]>0){
                    int w=j+a[i];
                    if(w>200)w=200;
                    if(dp[i+1][w])f=1;
                    else fz=0;
                }
                if(b[i]>0){
                    int w=j-b[i];
                    if(w<0)w=0;
                    if(dp[i+1][w])f=1;
                    else fz=0;
                }
                if(c[i]==1){
                    int w=j-100;
                    w=-w;
                    w+=100;
                    if(dp[i+1][w])f=1;
                    else fz=0;
                }
                if(f&&fz)dp[i][j]=1;
            }
        }
    }
    if(dp[1][m+100]){
        puts("Bad Ending");
        return 0;
    }
    puts("Normal Ending");
    return 0;
}

File: 2018/9.9jisuanke_xuzhou/B.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.Arrays;

public class Main{
    static int MAXN=1000+10;
    static int a[]=new int[MAXN];
    static int b[]=new int[MAXN];
    static int c[]=new int[MAXN];
    static boolean dp[][] = new boolean[MAXN][220];
    static StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
    static PrintWriter out = new PrintWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args)throws Exception{
        int n=getInt();
        int m=getInt();
        int k=getInt();
        int l=getInt();
        for(int i=1;i<=n;i++){
            a[i]=getInt();
            b[i]=getInt();
            c[i]=getInt();
        }
        for(int i=1;i<=n+1;i++){
            for(int j=0;j<=200;j++){
                dp[i][j]=false;
            }
        }
        for(int i=k+100;i<=200;i++){
            dp[n+1][i]=true;
        }
        for(int i=n;i>=1;i--){
            if(i%2==1){
                for(int j=0;j<=200;j++){
                    boolean f=false;
                    if(a[i]>0){
                        int w=j+a[i];
                        if(w>200)w=200;
                        if(dp[i+1][w])f=true;
                    }
                    if(b[i]>0){
                        int w=j-b[i];
                        if(w<0)w=0;
                        if(dp[i+1][w])f=true;
                    }
                    if(c[i]==1){
                        int w=j-100;
                        w=-w;
                        w+=100;
                        if(dp[i+1][w])f=true;
                    }
                    if(f)dp[i][j]=true;
                }
            }else{
                for(int j=0;j<=200;j++){
                    boolean f=false;
                    boolean fz=true;
                    if(a[i]>0){
                        int w=j+a[i];
                        if(w>200)w=200;
                        if(dp[i+1][w])f=true;
                        else fz=false;
                    }
                    if(b[i]>0){
                        int w=j-b[i];
                        if(w<0)w=0;
                        if(dp[i+1][w])f=true;
                        else fz=false;
                    }
                    if(c[i]==1){
                        int w=j-100;
                        w=-w;
                        w+=100;
                        if(dp[i+1][w])f=true;
                        else fz=false;
                    }
                    if(f&&fz)dp[i][j]=true;
                }
            }
        }
        if(dp[1][m+100]){
            out.println("Good Ending");
            out.flush();
            return;
        }
        for(int i=1;i<=n+1;i++){
            for(int j=0;j<=200;j++){
                dp[i][j]=false;
            }
        }
        for(int i=0;i<=l+100;i++){
            dp[n+1][i]=true;
        }
        for(int i=n;i>=1;i--){
            if(i%2==0){
                for(int j=0;j<=200;j++){
                    boolean f=false;
                    if(a[i]>0){
                        int w=j+a[i];
                        if(w>200)w=200;
                        if(dp[i+1][w])f=true;
                    }
                    if(b[i]>0){
                        int w=j-b[i];
                        if(w<0)w=0;
                        if(dp[i+1][w])f=true;
                    }
                    if(c[i]==1){
                        int w=j-100;
                        w=-w;
                        w+=100;
                        if(dp[i+1][w])f=true;
                    }
                    if(f)dp[i][j]=true;
                }
            }else{
                for(int j=0;j<=200;j++){
                    boolean f=false;
                    boolean fz=true;
                    if(a[i]>0){
                        int w=j+a[i];
                        if(w>200)w=200;
                        if(dp[i+1][w])f=true;
                        else fz=false;
                    }
                    if(b[i]>0){
                        int w=j-b[i];
                        if(w<0)w=0;
                        if(dp[i+1][w])f=true;
                        else fz=false;
                    }
                    if(c[i]==1){
                        int w=j-100;
                        w=-w;
                        w+=100;
                        if(dp[i+1][w])f=true;
                        else fz=false;
                    }
                    if(f&&fz)dp[i][j]=true;
                }
            }
        }
        if(dp[1][m+100]){
            out.println("Bad Ending");
            out.flush();
            return;
        }
        out.println("Normal Ending");
        out.flush();
    }
    static int getInt()throws IOException{
        in.nextToken();
        return (int) in.nval;
    }
}






